### ISSUE TITLE
cakephp 3 "date" pull-down panel, date field defined as DEFAULT NULL

### ISSUE THREAD

**Issue by renardrouge** [cakephp/cakephp#5613]
I have a "cake bake" generated form with a field defined in the database table as "date DEFAULT NULL".  
The date field is not a "required" field.  The resulting form is produced with pull-down tabs, sitting on the current UTC date.  
When I submit the form, the result seems to be the current UTC date and time . (for example, 1/9/15, 12:00 AM)  
I would like this field to remain NULL if not "required".  
How can I keep this field NULL when not selected, and when selected, is it possible to store the result in "date" format rather than "datetime" format ?

**Comment by markstory** [comment]
What version of CakePHP are you using?

**Comment by renardrouge** [comment]
I am using cakePHP 3.0.0-RC1 with the latest (about 2 hours ago) "composer.phar update --dev"

**Comment by renardrouge** [comment]
![null_date](https://cloud.githubusercontent.com/assets/10366395/5674464/d3ba530a-9778-11e4-9e97-c8356f23200a.JPG)   here is a partial screen shot

**Comment by markstory** [comment]
The hour components are being added as PHP does not have a 'date' type. It only has a datetime. Because Date columns do not store a time component, the ORM will set the time to midnight.

You should be able to leave the inputs empty to get a null value in the database though.

**Comment by renardrouge** [comment]
the database contains "2015-01-09" in the date field.

**Comment by renardrouge** [comment]
![null_date_2](https://cloud.githubusercontent.com/assets/10366395/5674941/7408d6ee-9781-11e4-8be4-5e5bdc17f82c.JPG)    this is a screenshot of the input form when it first appears.

**Comment by ADmad** [comment]
@markstory I believe the issue is that he expects the selects to be "empty" instead of having current date selected since the default value in db is null.

**Comment by renardrouge** [comment]
correct !

**Comment by markstory** [comment]
Oh. Have you tried setting `empty => true` when generating your input? This [was a recent change to bake](https://github.com/cakephp/bake/pull/25) for select boxes, it could also be applied to date inputs I guess.

**Comment by renardrouge** [comment]
I tried the following.   

echo $this->Form->input('date_of_birth', array('empty' => true));  
  
Default of today's UTC date still shows.

**Comment by markstory** [comment]
Defaulting to today is how the date inputs are intended to work. You can use the `default` option to set a default empty value.

```php
echo $this->Form->input('date_of_birth', ['empty' => true, 'default' => '']);
```

Will do what you need.

**Comment by renardrouge** [comment]
Almost there Mark !  I get the blank date field with your suggested change.   However, the Submit results in a flash message "The user could not be saved. Please, try again." and the "Date of Birth" year field is highlighted in red, with a "00" inserted. with "The provided value is invalid" appearing below the field.  I am assuming that no changes have been made in this area since my update 12 days ago, correct ?                                              Here is a screen shot.    
![null_date_3](https://cloud.githubusercontent.com/assets/10366395/5830533/ce5e9e9a-a0ef-11e4-998c-847de2aa7001.JPG)

**Comment by renardrouge** [comment]
Also, before submission. the date pull down ranges from 2010 to 2020. When the "validation error" is reported, the pull down ranges from 00 to 2020.

**Comment by markstory** [comment]
Do you have validation for the date of birth field? What does that validation look like?

**Comment by renardrouge** [comment]
Good morning Mark !
The code is almost completely generated by "cake bake all users".
This morning at 10:30 a.m. I did a "php composer.phar update --dev" and did a "cake bake all users" into a completely new application.  I then modified only this line in .../Users/add.ctp.
echo $this->Form->input('date_of_birth');
The modification included
                				, array('empty' => true,
           					'default' => '',
           					'novalidate' => true));
I tried first without the 'novalidate' and the same error as reported previously occurs.
The 'novalidate' addition made no difference.

Here is my database definition.
CREATE DATABASE datebug;
USE datebug;
CREATE TABLE IF NOT EXISTS users (	
  id			int(10) UNSIGNED NOT NULL AUTO_INCREMENT,
  username		varchar(254) NOT NULL,	
  user_email		varchar(254) NOT NULL,
  date_of_birth		date        DEFAULT NULL,
  UNIQUE KEY (username),
  KEY (user_email),
  PRIMARY KEY (id)
)

Here is my final screenshot.
![null_date_4](https://cloud.githubusercontent.com/assets/10366395/5839657/ac76738e-a15d-11e4-8e82-4658933f329c.JPG)

Regards
Lorne

**Comment by renardrouge** [comment]
From very early in this problem (before I updated to most current dev code) this partial stack trace suggests that "cake bake" is generating code to validate.

    Cake\Validation\Validation::_check() - CORE\src\Validation\Validation.php, line 763  
    Cake\Validation\Validation::date() - CORE\src\Validation\Validation.php, line 346  
    Cake\Validation\RulesProvider::__call() - CORE\src\Validation\RulesProvider.php, line 53  
    Cake\Validation\RulesProvider::date() - CORE\src\Validation\ValidationRule.php, line 138  

I don't mind making whatever alterations may be required to get this working, but I come back to my original position.

"cake bake" code on a field defined as DEFAULT NULL should work directly out of the "cake oven".
i.e. no input to a DEFAULT NULL field, no validation, no entry to the database.

Regards
Lorne

**Comment by renardrouge** [comment]
Here is the complete stack trace back in time (2015-01-08).  I note that some of the code has changed in the validation area since this trace was taken. 

    preg_match - [internal], line ??    
    Cake\Validation\Validation::_check() - CORE\src\Validation\Validation.php, line 763  
    Cake\Validation\Validation::date() - CORE\src\Validation\Validation.php, line 346  
    Cake\Validation\RulesProvider::__call() - CORE\src\Validation\RulesProvider.php, line 53  
    Cake\Validation\RulesProvider::date() - CORE\src\Validation\ValidationRule.php, line 138  
    Cake\Validation\ValidationRule::process() - CORE\src\Validation\ValidationRule.php, line 138  
    Cake\Validation\Validator::_processRules() - CORE\src\Validation\Validator.php, line 554  
    Cake\Validation\Validator::errors() - CORE\src\Validation\Validator.php, line 128  
    Cake\ORM\Marshaller::_validate() - CORE\src\ORM\Marshaller.php, line 185  
    Cake\ORM\Marshaller::merge() - CORE\src\ORM\Marshaller.php, line 341  
    Cake\ORM\Table::patchEntity() - CORE\src\ORM\Table.php, line 1899  
    App\Controller\UsersController::add() - APP/Controller\UsersController.php, line 48  
    Cake\Controller\Controller::invokeAction() - CORE\src\Controller\Controller.php, line 406  
    Cake\Routing\Dispatcher::_invoke() - CORE\src\Routing\Dispatcher.php, line 114  
    Cake\Routing\Dispatcher::dispatch() - CORE\src\Routing\Dispatcher.php, line 87

**Comment by renardrouge** [comment]
Starting at line 1802 in /src/ORM.Table.php I see these comments, which are contrary to the behaviour that I currently see.

     * By default, the data is validated before being passed to the new entity. In
     * the case of invalid fields, those will not be present in the resulting object.
     * The `validate` option can be used to disable validation on the passed data:
     *
     * ```
     * $article = $this->Articles->newEntity(
     *   $this->request->data(),
     *   ['validate' => false]
     * );
     * ```
     *
     * You can also pass the name of the validator to use in the `validate` option.
     * If `null` is passed to the first param of this function, no validation will
     * be performed.

**Comment by markstory** [comment]
Alright, I have a good idea of what is going on now.

**Comment by markstory** [comment]
Fixed with #5723
